test/jdk/sun/security/tools/keytool/GenKeyPairSigner.java

```diff
@@ -0,0 +1,326 @@
+         * along with -signer option.
+         */
+        System.out.println("Testing the signer alias that is stored in the JKS keystore");
+        ktjks("-genkeypair -keyalg RSA -keysize 1024 -alias ca -dname CN=CA -ext bc:c",
+                "ksjks", "cakeypass");
+
+        System.out.println("Generating an DSA cert with -signer and -signerkeypass options");
+        SecurityTools.keytool("-keystore ksjks -storepass changeit -storetype jks " +
+                "-genkeypair -keyalg DSA -keysize 1024 -alias ca1 -dname CN=CA1 " +
+                "-keypass ca1keypass -signer ca -signerkeypass cakeypass")
+                .shouldContain("Generating 1,024 bit DSA key pair and a certificate (SHA256withRSA) issued by <ca> with a validity of 90 days")
+                .shouldContain("for: CN=CA1")
+                .shouldContain("The generated certificate #1 of 2 uses a 1024-bit DSA key which is considered a security risk")
+                .shouldContain("The generated certificate #2 of 2 uses a 1024-bit RSA key which is considered a security risk")
+                .shouldHaveExitValue(0);
+
+        System.out.println("Generating an XDH cert with -signer and -signerkeypass options");
+        SecurityTools.keytool("-keystore ksjks -storepass changeit -storetype jks " +
+                "-genkeypair -keyalg XDH -alias e1 -dname CN=E1 " +
+                "-keypass e1keypass -signer ca1 -signerkeypass ca1keypass")
+                .shouldContain("Generating 255 bit XDH key pair and a certificate (SHA256withDSA) issued by <ca1> with a validity of 90 days")
+                .shouldContain("for: CN=E1")
+                .shouldContain("The generated certificate #2 of 3 uses a 1024-bit DSA key which is considered a security risk")
+                .shouldContain("The generated certificate #3 of 3 uses a 1024-bit RSA key which is considered a security risk")
+                .shouldHaveExitValue(0);
+
+        // examine the resulting cert
+        kstore = KeyStore.getInstance(new File("ksjks"), "changeit".toCharArray());
+        cert = (X509Certificate)kstore.getCertificate("e1");
+
+        Certificate[] certChain = kstore.getCertificateChain("e1");
+        if (certChain.length != 3) {
+            System.out.println("Generated cert chain is in error");
+            System.exit(1);
+        }
+
+        sigName = cert.getSigAlgName();
+        if (sigName != "SHA256withDSA") {
+            System.out.println("Signature algorithm name is in error");
+            System.exit(1);
+        }
+
+        pKey = cert.getPublicKey();
+        keyLen = KeyUtil.getKeySize(pKey);
+        if (keyLen != 255) {
+            System.out.println("Key size is in error");
+            System.exit(1);
+        }
+
+        pKeyAlg = pKey.getAlgorithm();
+        if (pKeyAlg != "XDH") {
+            System.out.println("Subject Public Key Algorithm is in error");
+            System.exit(1);
+        }
+
```

Comment:
Since you are here, you can check if the new entry is indeed protected by the new key password.

Reply:
In testSignerJKS() has made sure that the new entry created with new key password can *only* be accessed when a correct key password is provided in order to retrieve the corresponding signer’s private key.

Reply:
The new entry is protected by the new key password is an existing function, and its testing should have been covered.

Reply:
OK.